src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -121,47 +139,105 @@ private void ParseMarkupElement(in SyntaxListBuilder<RazorSyntaxNode> builder)
             if (!NextIs(SyntaxKind.ForwardSlash))
             {
                 // Parsing a start tag
+                var tagStart = CurrentStart;
                 var startTag = ParseStartTag(out var tagName, out var tagMode);
-                builder.Add(startTag);
-
                 if (tagMode == MarkupTagMode.Script)
                 {
-                    ParseJavascriptAndEndScriptTag(builder);
+                    ParseJavascriptAndEndScriptTag(builder, startTag);
+                    return;
+                }
+                else if (tagMode == MarkupTagMode.SelfClosing || tagMode == MarkupTagMode.Invalid)
+                {
+                    // For cases like <foo /> or invalid cases like |<|<p>
+                    var element = SyntaxFactory.MarkupElement(startTag, SyntaxList<RazorSyntaxNode>.Empty, null);
+                    builder.Add(element);
+                    return;
+                }
+                else
+                {
+                    // This is a normal start tag. We need to keep track of it.
+                    var tracker = new TagTracker(tagName, startTag, tagStart, builder);
+                    _tagTracker.Push(tracker);
+                    builder.Clear();
+                    return;
+                }
+            }
+            else
+            {
+                // Parsing an end tag.
+                var endTag = ParseEndTag(out var endTagName);
+                if (endTagName != null && string.Equals(CurrentStartTagName, endTagName, StringComparison.OrdinalIgnoreCase))
+                {
+                    // Happy path. Found a matching start tag. Create the element and reset the builder.
+                    var tracker = _tagTracker.Pop();
+                    var element = SyntaxFactory.MarkupElement(tracker.StartTag, builder.ToList(), endTag);
+                    builder.Clear();
```

Comment:
Just a nit, would probably make sense to add a `Builder.Consume()` or something like it to indicate `.ToList() -> .Clear()`. From my understanding of it, it never makes sense to call `.ToList()` and pass it onto something and not clear it afterwards.

Reply:
Hmm I like that. I'll try it